staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery.go

```diff
@@ -109,55 +180,148 @@ func (r *groupDiscoveryHandler) ServeHTTP(w http.ResponseWriter, req *http.Reque
+			continue
+		}
+
+		for _, v := range crd.Spec.Versions {
+			if !v.Served {
+				continue
+			}
+			// If there is any Served version, that means the group should show up in discovery
+			foundGroup = true
+
+			// HACK: support the case when we add core resources through CRDs (KCP scenario)
+			groupVersion := crd.Spec.Group + "/" + v.Name
+			if crd.Spec.Group == "" {
+				groupVersion = v.Name
+			}
+
+			gv := metav1.GroupVersion{Group: crd.Spec.Group, Version: v.Name}
+
+			if !versionsForDiscoveryMap[gv] {
+				versionsForDiscoveryMap[gv] = true
+				apiVersionsForDiscovery = append(apiVersionsForDiscovery, metav1.GroupVersionForDiscovery{
+					GroupVersion: groupVersion,
+					Version:      v.Name,
+				})
+			}
+		}
+	}
+
+	sortGroupDiscoveryByKubeAwareVersion(apiVersionsForDiscovery)
 
-	discovery, ok := r.getDiscovery(clusterName, pathParts[1])
-	if !ok {
+	// HACK: if we are adding resources in legacy scheme group through CRDs (KCP scenario)
+	// then do not expose the CRD `APIResource`s in their own CRD-related group`,
+	// But instead add them in the existing legacy schema group
+	// if genericcontrolplanescheme.Scheme.IsGroupRegistered(clusterGroupVersion.Group) {
+	// 	if !foundGroup || !foundVersion {
+	// 		delete(discovery.ContributedResources, clusterGroupVersion)
+	// 	}
+
+	// 	discovery.ContributedResources[clusterGroupVersion] = resourceListerFunc
+	// 	return nil
+	// }
+
+	if !foundGroup {
 		r.delegate.ServeHTTP(w, req)
 		return
 	}
 
-	discovery.ServeHTTP(w, req)
-}
+	apiGroup := metav1.APIGroup{
+		Name:     requestedGroup,
+		Versions: apiVersionsForDiscovery,
+		// the preferred versions for a group is the first item in
+		// apiVersionsForDiscovery after it put in the right ordered
+		PreferredVersion: apiVersionsForDiscovery[0],
+	}
 
-func (r *groupDiscoveryHandler) getDiscovery(clusterName, group string) (*discovery.APIGroupHandler, bool) {
-	r.discoveryLock.RLock()
-	defer r.discoveryLock.RUnlock()
+	discovery.NewAPIGroupHandler(Codecs, apiGroup).ServeHTTP(w, req)
+}
 
-	clusterDiscovery, clusterExists := r.discovery[clusterName]
-	if !clusterExists {
-		return nil, false
-	}
-	ret, ok := clusterDiscovery[group]
-	return ret, ok
+type rootDiscoveryHandler struct {
+	crdLister listers.CustomResourceDefinitionLister
+	delegate  http.Handler
 }
 
-func (r *groupDiscoveryHandler) setDiscovery(clusterName, group string, discoveryHandler *discovery.APIGroupHandler) {
-	r.discoveryLock.Lock()
-	defer r.discoveryLock.Unlock()
+func (r *rootDiscoveryHandler) Groups(ctx context.Context, req *http.Request) ([]metav1.APIGroup, error) {
+	apiVersionsForDiscovery := map[string][]metav1.GroupVersionForDiscovery{}
+	versionsForDiscoveryMap := map[string]map[metav1.GroupVersion]bool{}
 
-	if _, clusterExists := r.discovery[clusterName]; !clusterExists {
-		r.discovery[clusterName] = map[string]*discovery.APIGroupHandler{}
+	crds, err := r.crdLister.ListWithContext(ctx, labels.Everything())
```

Comment:
Yeah, I had that same thought, but didn't want to bother while I was hacking away. This code came from a controller, and it was unified because it set up both group and version discovery at the same time. But I ended up splitting it out into live handlers for each, and they're building different data sets. Maybe worthy of a TODO for now?